src/modifier.ts

```diff
@@ -51,126 +69,134 @@ export class Modifier extends Element {
     this.x_shift = 0;
     this.y_shift = 0;
     this.spacingFromNextModifier = 0;
   }
 
   // Called when position changes
-  reset() {
+  reset(): void {
     // do nothing
   }
 
   // Every modifier has a category. The `ModifierContext` uses this to determine
   // the type and order of the modifiers.
-  getCategory() {
+  getCategory(): string {
     return Modifier.CATEGORY;
   }
 
   // Get and set modifier widths.
-  getWidth() {
+  getWidth(): number {
     return this.width;
   }
-  setWidth(width) {
+
+  setWidth(width: number): this {
     this.width = width;
     return this;
   }
 
   // Get and set attached note (`StaveNote`, `TabNote`, etc.)
-  getNote() {
+  getNote(): Note {
+    if (!this.note) throw new Vex.RERR('NoNote', 'Modifier has no note.');
     return this.note;
   }
-  setNote(note) {
+
+  setNote(note: Note): this {
     this.note = note;
     return this;
   }
 
   // Get and set note index, which is a specific note in a chord.
-  getIndex() {
+  getIndex(): number {
     return this.index;
   }
-  setIndex(index) {
+
+  setIndex(index: number): this {
     this.index = index;
     return this;
   }
 
   // Every modifier must be part of a `ModifierContext`.
-  getModifierContext() {
+  getModifierContext(): ModifierContext | undefined {
     return this.modifier_context;
   }
-  setModifierContext(c) {
+
+  setModifierContext(c: ModifierContext): this {
     this.modifier_context = c;
     return this;
   }
 
   // Get and set articulation position.
-  getPosition() {
+  getPosition(): number {
     return this.position;
   }
-  setPosition(position) {
+
+  setPosition(position: number): this {
     this.position = typeof position === 'string' ? Modifier.PositionString[position] : position;
     this.reset();
     return this;
   }
 
   // Set the `text_line` for the modifier.
-  setTextLine(line) {
+  setTextLine(line: number): this {
     this.text_line = line;
     return this;
   }
 
   // Shift modifier down `y` pixels. Negative values shift up.
-  setYShift(y) {
+  setYShift(y: number): this {
     this.y_shift = y;
     return this;
   }
 
-  setSpacingFromNextModifier(x) {
+  setSpacingFromNextModifier(x: number): void {
     this.spacingFromNextModifier = x;
   }
 
-  getSpacingFromNextModifier() {
+  getSpacingFromNextModifier(): number {
     return this.spacingFromNextModifier;
   }
 
   // Shift modifier `x` pixels in the direction of the modifier. Negative values
   // shift reverse.
-  setXShift(x) {
+  setXShift(x: number): this {
     this.x_shift = 0;
     if (this.position === Modifier.Position.LEFT) {
       this.x_shift -= x;
     } else {
       this.x_shift += x;
     }
+    return this;
   }
-  getXShift() {
+
+  getXShift(): number {
     return this.x_shift;
   }
 
   // Render the modifier onto the canvas.
-  draw() {
-    this.checkContext();
+  draw(): void {
     throw new Vex.RERR('MethodNotImplemented', 'draw() not implemented for this modifier.');
```

Comment:
Not sure why `checkContext` was removed here.

Reply:
To reduce size of code, I thought it makes no sense checking that the context is there before throwing a `RERR`.
I also spent some time asking myself why we would do that.

Reply:
This will cause a crash for logical elements that don't render anything, but just reserve space. So they will have a context, but not draw anything. In general, better not to change logic in these commits -- it's very hard to predict the downstream effects.

Reply:
ok